File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    ArrayList<Integer>[] map;    
    int cnt;
    
    public int solution(int n, int[][] wires) {
        init(n, wires);
        int answer = 101;
        for(int i=0; i<n-1; i++){
            int a = wires[i][0];
            int b = wires[i][1];
            
            map[a].add(b);
            map[b].add(a);
        }
        
        for(int i=0; i<n-1; i++){
            int a = wires[i][0];
            int b = wires[i][1];
            
            int v1 = cut(a, b);
            int v2 = cut(b, a);
            
            int abs = Math.abs(v1-v2);
            answer = Math.min(abs, answer);
        }
        
        return answer;
    }
    
    public int cut(int start, int cuting){
        boolean[] visited = new boolean[cnt+1];
        visited[cuting] = true;
        visited[start] = true;
        
        Queue<Integer> que = new ArrayDeque<>();
        que.add(start);
        int cnt = 0;
        while(!que.isEmpty()){
            int p = que.poll();
            cnt++;
            for(int next : map[p]){
                if(visited[next]){
                    continue;
                }
                visited[next] = true; 
                que.add(next);
                
            }            
        }
        
        return cnt;
        
    }
    
    public void init(int n, int[][] wires){
        map = new ArrayList[n+1];
        for(int i=1; i<=n; i++){
            map[i] = new ArrayList<>();
        }
        cnt = n;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    answer = 1000
    
    graph = {}
    
    for s,e in wires:
        if s in graph:
            graph[s].append(e)
        else:
            graph[s]=[e]
        if e in graph:
            graph[e].append(s)
        else:
            graph[e]=[s]
    
    print(graph)
    
    for s1, s2 in wires:
        visited = [0]*101
        visited[s1] = 1
        visited[s2] = 1
        a = bfs(s1, s2,graph, visited)
        b =bfs(s2, s1,graph, visited)
        answer = min(answer, abs(a - b))  
    
    return answer

def bfs(node, cut, graph, visited):
    
    count = 1
    
    que = [node]
    
    while(len(que) > 0):
        
        n = que.pop()
        
        for i in graph[n]:
            if not visited[i]:
                visited[i] = 1
                count += 1
                que.append(i)
    
    
    return count